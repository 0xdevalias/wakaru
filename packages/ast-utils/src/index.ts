export { isFunctionExpression } from './isFunctionExpression'
export { isTopLevel } from './isTopLevel'
export { renameFunctionParameters } from './renameFunctionParameters'
export { getTopLevelStatements } from './program'
export { pruneComments } from './pruneComments'
export { wrapAstTransformation } from './wrapAstTransformation'

export * from './comments'
export * from './identifier'
export * from './imports'
export * from './isIIFE'
export * from './exports'
export * from './scope'
export * from './types'
export * from './object'
export * from './position'
export * from './reference'
