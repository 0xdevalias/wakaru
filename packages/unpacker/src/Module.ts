import type { Collection } from 'jscodeshift'

export class Module {
    id: string | number
    ast: Collection
    isEntry: boolean

    constructor(id: string | number, ast: Collection, isEntry = false) {
        this.id = id
        this.ast = ast
        this.isEntry = isEntry
    }
}
