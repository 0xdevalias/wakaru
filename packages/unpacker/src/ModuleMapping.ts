export type ModuleMapping = Record<number | string, string>
