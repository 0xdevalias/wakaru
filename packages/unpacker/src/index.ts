export type { Module } from './Module'
export type { ModuleMapping } from './ModuleMapping'

export * from './unpack'
