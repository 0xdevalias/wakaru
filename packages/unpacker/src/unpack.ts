import jscodeshift from 'jscodeshift'

// @ts-expect-error - no types
import getParser from 'jscodeshift/src/getParser'
import { getModulesFromBrowserify } from './extractors/browserify'
import { getModulesFromWebpack } from './extractors/webpack'
import type { Module } from './Module'
import type { ModuleMapping } from './ModuleMapping'

/**
 * Unpacks the given source code from supported bundlers.
 */

export function unpack(sourceCode: string): {
    modules: Module[]
    moduleIdMapping: ModuleMapping
} | null {
    const parser = getParser()
    const j = jscodeshift.withParser(parser)
    const root = j(sourceCode)

    const result = getModulesFromWebpack(j, root)
        || getModulesFromBrowserify(j, root)

    if (!result) {
        console.error('Failed to locate modules')
        return null
    }

    const { modules, moduleIdMapping } = result

    return {
        modules: [...modules],
        moduleIdMapping,
    }
}
