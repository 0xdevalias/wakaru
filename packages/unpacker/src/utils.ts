import type { ASTPath, ArrowFunctionExpression, Collection, ExpressionStatement, FunctionExpression, JSCodeshift, Node, Statement } from 'jscodeshift'
import prettier from 'prettier/standalone'
import babelParser from 'prettier/parser-babel'

export function isTopLevel(j: JSCodeshift, node: ASTPath<Node>): boolean {
    return j.Program.check(node.parentPath.node)
}

export function pruneComments(j: JSCodeshift, collection: Collection<any>): void {
    // @ts-expect-error - Comment type is wrong
    collection.find(j.Comment).forEach(path => path.prune())
}

export function renameFunctionParameters(j: JSCodeshift, node: FunctionExpression | ArrowFunctionExpression, parameters: string[]): void {
    node.params.forEach((param, index) => {
        if (param.type === 'Identifier') {
            j(node)
                .find(j.Identifier, { name: param.name })
                .filter(path => path.scope.node === node)
                .forEach((path) => {
                    path.node.name = parameters[index]
                })
        }
    })
}

export function isIIFE(node: Statement): node is ExpressionStatement {
    if (node.type !== 'ExpressionStatement') return false
    const expression = (node as ExpressionStatement).expression
    if (expression.type !== 'CallExpression') return false
    const callee = expression.callee
    return callee.type === 'FunctionExpression'
        || callee.type === 'ArrowFunctionExpression'
}

export function prettierFormat(code: string) {
    return prettier.format(code, {
        parser: 'babel',
        plugins: [babelParser],
    })
}
