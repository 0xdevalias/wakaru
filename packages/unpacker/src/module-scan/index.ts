import { ExportManager } from '@wakaru/ast-utils/exports'
import { ImportManager } from '@wakaru/ast-utils/imports'
import { postScanBabelRuntime, scanBabelRuntime } from './babel-runtime'
import type { Module } from '../Module'
import type { Collection, JSCodeshift } from 'jscodeshift'

export function scanImports(j: JSCodeshift, module: Module & { root: Collection }) {
    const importManager = new ImportManager()
    importManager.collectEsModuleImport(j, module.root)
    importManager.collectCommonJsImport(j, module.root)
    module.import = importManager.getModuleImports()
}

export function scanExports(j: JSCodeshift, module: Module & { root: Collection }) {
    const exportManager = new ExportManager()
    exportManager.collectEsModuleExport(j, module.root)
    exportManager.collectCommonJsExport(j, module.root)
    module.export = exportManager.toJSON()
}

export function scanRuntime(j: JSCodeshift, module: Module & { root: Collection }) {
    scanBabelRuntime(j, module)
}

export function postScanRuntime(j: JSCodeshift, modules: (Module & { root: Collection })[]) {
    postScanBabelRuntime(j, modules)
}
