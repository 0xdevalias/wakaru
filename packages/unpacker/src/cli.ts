#!/usr/bin/env node

/* eslint-disable no-console */
import * as path from 'node:path'
import process, { hrtime } from 'node:process'
import fsa from 'fs-extra'
import * as globby from 'globby'
import c from 'picocolors'
import yargs from 'yargs'
import { hideBin } from 'yargs/helpers'
import { version } from '../package.json'
import { unpack } from './unpack'

type LogLevel = 'error' | 'warn' | 'info' | 'debug' | 'silent'

// eslint-disable-next-line no-unused-expressions
yargs(hideBin(process.argv))
    .scriptName('@wakaru/unpacker')

    .help()
    .showHelpOnFail(true)
    .alias('h', 'help')

    .version('version', version)
    .alias('v', 'version')

    .option('log-level', {
        type: 'string',
        default: 'info',
        choices: ['error', 'warn', 'info', 'debug', 'silent'],
        describe: 'change the level of logging for the CLI.',
    })

    .usage('Usage: $0 <files...> [options]')
    .command(
        '* <files...>',
        'Unminify your bundled code',
        args => args
            .option('output', {
                alias: 'o',
                describe: 'specify the output directory (default: out/)',
                type: 'string',
                default: 'out/',
            })
            .option('force', {
                alias: 'f',
                describe: 'force overwrite output directory',
                type: 'boolean',
                default: false,
            })
            .positional('files', {
                describe: 'File paths to process (supports glob patterns)',
                type: 'string',
                array: true,
            })
            .help(),
        async (args) => {
            await unpacker(
                args.files ?? [],
                args.output,
                args.force,
                args.logLevel as LogLevel,
            )
        },
    )
    .argv

async function unpacker(
    paths: string[],
    output: string,
    force: boolean,
    logLevel: LogLevel,
) {
    const cwd = process.cwd()
    const globbyPaths = paths
        .map(p => path.normalize(p))
        .map(p => p.replace(/\\/g, '/'))
    const resolvedPaths = await globby.default(globbyPaths, {
        cwd,
        absolute: true,
        ignore: [path.join(cwd, '**/node_modules/**')],
    })

    // Check if any paths are outside of the current working directory
    for (const p of resolvedPaths) {
        if (!isPathInside(cwd, p)) {
            throw new Error(`File path ${c.green(path.relative(cwd, p))} is outside of the current working directory. This is not allowed.`)
        }
    }

    const outputDir = path.resolve(cwd, output)

    if (await fsa.exists(outputDir)) {
        if (!force) {
            throw new Error(`Output directory already exists at ${c.green(path.relative(cwd, outputDir))}. Pass ${c.yellow('--force')} to overwrite.`)
        }
    }
    await fsa.ensureDir(outputDir)

    for (const p of resolvedPaths) {
        const start = hrtime()

        const source = await fsa.readFile(p, 'utf-8')
        const { modules, moduleIdMapping } = unpack(source)

        const folder = resolvedPaths.length > 1 ? path.basename(p, path.extname(p)) : ''
        const outputFolder = path.join(outputDir, folder)
        for (const mod of modules) {
            const filename = moduleIdMapping[mod.id] ?? `module-${mod.id}.js`
            const outputPath = path.join(outputFolder, filename)
            await fsa.ensureDir(path.dirname(outputPath))
            await fsa.writeFile(outputPath, mod.code, 'utf-8')
        }

        if (logLevel !== 'silent') {
            const end = hrtime(start)
            const elapsed = end[0] * 1e9 + end[1]
            const formattedElapsed = (elapsed / 1e6).toLocaleString('en-US', { maximumFractionDigits: 1 })
            console.log(`Generated ${c.green(modules.length)} modules from ${c.green(path.relative(cwd, p))} to ${c.green(path.relative(cwd, outputFolder))} ${c.dim(`(${formattedElapsed}ms)`)}`)
        }
    }
}

/**
 * Check if base path contains target path
 */
function isPathInside(base: string, target: string): boolean {
    const relative = path.relative(base, target)
    return !relative.startsWith('..') && !path.isAbsolute(relative)
}
