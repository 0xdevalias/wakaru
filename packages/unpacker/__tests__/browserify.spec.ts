import { readFile } from 'node:fs/promises'
import { describe, expect, it } from 'vitest'
import { unpack } from '../src'

describe('Browserify', () => {
    it('testcases/browserify', async () => {
        const source = await readFile('../../testcases/browserify/dist/index.js', 'utf8')
        const result = unpack(source)
        if (!result) throw new Error('Failed to unpack')

        expect(result.moduleIdMapping).toMatchSnapshot()

        expect(result.modules.length).toBe(4)

        const modules = [...result.modules.values()]
            .map(({ id, isEntry, code }) => ({ id, isEntry, code }))
        expect(modules).toMatchSnapshot()
    })
})
