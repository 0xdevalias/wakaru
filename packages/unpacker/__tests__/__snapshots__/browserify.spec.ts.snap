// Vitest Snapshot v1

exports[`Browserify > testcases/browserify 1`] = `
Map {
  1 => "is-array",
  2 => "./calculator",
  3 => "./greeting",
}
`;

exports[`Browserify > testcases/browserify 2`] = `
[
  {
    "code": "/**
 * isArray
 */

var isArray = Array.isArray;

/**
 * toString
 */

var str = Object.prototype.toString;

/**
 * Whether or not the given \`val\`
 * is an array.
 *
 * example:
 *
 *        isArray([]);
 *        // > true
 *        isArray(arguments);
 *        // > false
 *        isArray('');
 *        // > false
 *
 * @param {mixed} val
 * @return {bool}
 */

module.exports = isArray || function (val) {
  return !! val && '[object Array]' == str.call(val);
};",
    "id": 1,
    "isEntry": false,
  },
  {
    "code": "const isArray = require('is-array')

const pi = 3.14

module.exports.pi = pi

class Calculator {
    add(a, b) {
        return a + b
    }

    sub(a, b) {
        return a - b
    }

    sum(arr) {
        if (!isArray(arr)) throw new Error('Argument must be an array')
        return arr.reduce((a, b) => a + b, 0)
    }

    pi() {
        return pi
    }
}

module.exports.Calculator = Calculator

module.exports.mul = (a, b) => a * b

function div(a, b) {
    return a / b
}

module.exports.div = div

const isArr = arr => isArray(arr)

module.exports.isArr = isArr",
    "id": 2,
    "isEntry": false,
  },
  {
    "code": "const { add, pi } = require('./calculator')

class Greeting {
    constructor(name) {
        this.name = name
    }

    sayHello() {
        console.log(\`Hello \${this.name}. Your lucky number is \${add(pi, 2)}\`)
    }
}

module.exports = Greeting",
    "id": 3,
    "isEntry": false,
  },
  {
    "code": "const { Calculator, div, mul, pi } = require('./calculator')
const Greeting = require('./greeting')

const cal = new Calculator()
console.log(cal.add(pi, 2))
console.log(mul(1, 2))
console.log(div(1, 2))

const greeting = new Greeting('John')
greeting.sayHello()",
    "id": 4,
    "isEntry": true,
  },
]
`;
